Deduplicate messages when deserializing

rasmusfaber opened the pull request:
## This PR contains:
- [ ] New features
- [ ] Changes to dev-tools e.g. CI config / github tooling
- [ ] Docs
- [ ] Bug fixes
- [X] Code refactor

### What is the current behavior? (You can also link to an open issue here)

While deserializing the sample logs, repeated references to the same logical message are not preserved. Each occurrence is turned into a new model instance, so messages that referred to the same entity during serialization end up as different instances.

This causes severe memory blowup for long sample runs since the length of the messages in the model events typically grow linearly (and total memory usage thus grows quadratically).

### What is the new behavior?

During deserialization we cache messages by id and reuse object instances.


### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)

### Other information:

See also Slack thread.

jjallaire commented:
This is looking good! I wonder if we can go one step further and 

(a) Give the context a bit more typing (e.g. a TypedDict)
(b) Implement optional bypassing of resolving attachments in model api calls -- this will also substantially improve memory consumption for model events and it seems like it would be quite rare that you want to look at the raw model api text content (that data is essentially for the viewer for debugging/diagnostics).

So maybe `resolve_attachments` takes `bool | "core" | "debug"` (or something like that) and defaults to "core", which skips resolving attachments in the `.call` field of ModelEvent.

cc @sjawhar

rasmusfaber commented:
@jjallaire I added typing to the walk context, and added an attempt to have both `"core" `and `"full"` resolve_attachments. Isn't `"full" `a better name than `"debug"`? Are you okay with a breaking change by changing the default to `"core"`?

sjawhar commented on src/inspect_ai/model/_chat_message.py:
```diff
@@ -54,6 +55,20 @@ def model_post_init(self, __context: Any) -> None:
         if self.id is None and not is_deserializing:
             self.id = uuid()
 
+    @model_validator(mode="after")
+    def _after(self, info: ValidationInfo) -> "ChatMessageBase":
```
It's not clear to me how an after validator helps in this case, since the object will have already been deserialized. I think a `wrap` validator is more appropriate

rasmusfaber replied:
Yes, this is slightly wasteful, but I needed the parsed `ChatMessage` in order to do the equality check. (Just in case this is not clear: many eval log files have `ChatMessage`s with the same id, but different content. For example, the submit tool_call might be stripped in some places, but not in others).

How do you like this version instead? I pulled in the `frozendict` dependency, and it now uses a bit more memory for the cache. On the other hand we save creating a bunch of unnecessary `ChatMessag` instances.

jjallaire replied:
That sounds good to me.

sjawhar commented on src/inspect_ai/_cli/log.py:
```diff
@@ -129,15 +129,29 @@ def list_command(
 )
 @click.option(
     "--resolve-attachments",
+    "resolve_flag",
     type=bool,
     is_flag=True,
     default=False,
     help="Resolve attachments (duplicated content blocks) to their full content.",
 )
-def dump_command(path: str, header_only: bool, resolve_attachments: bool) -> None:
+@click.option(
+    "--resolve-attachments-mode",
+    type=click.Choice(["full", "core"]),
+    default="core",
+    help="How to resolve attachments.",
+)
```
See e.g. the `--stream` parameter I added for scoring/conversion, which defaults to `True` but can also be set to an integer

rasmusfaber replied:
How is this instead? This is a breaking change in that just `--resolve-attachments` is now no longer valid. I could also add a second flag-type `--resolve-attachments` parameter to keep the old behavior, but I think this might be slightly better to force users to select either `full` or `core.

sjawhar replied:
> This is a breaking change in that just `--resolve-attachments` is now no longer valid.

Why? Again, other commands have the functionality JJ and I have suggested. What's the blocker here?

sjawhar commented on src/inspect_ai/_util/constants.py:
```diff
@@ -44,4 +44,8 @@
+
+
+def get_deserializing_context() -> dict[str, Any]:
+    return {DESERIALIZING: True, MESSAGE_CACHE: {}}
```
This returns a fresh cache with each invocation, which seems like a potential footgun (honestly, though, so does the opposite)

jjallaire replied:
Yeah, I think it's fine and probably in fact best to bound the cache to the current deserialization context (which is a very top level thing)

sjawhar commented on =2.0.0:
What is this file?

sjawhar commented on src/inspect_ai/_cli/log.py:
```diff
@@ -164,9 +183,11 @@ def dump_command(path: str, header_only: bool, resolve_attachments: bool) -> Non
+    type=click.Choice(["full", "core"]),
+    flag_value="core",
+    is_flag=False,
+    default=None,
```
I notice that these args are different than [here](https://github.com/UKGovernmentBEIS/inspect_ai/pull/2552/files#diff-18829141d4859bd887124c0c522e91bf1046f3fc79241849f0af6d1074d73d9cR149-R150)

sjawhar commented:
You'd definitely finish it before I get the chance 😅

jjallaire commented:
> You'd definitely finish it before I get the chance 😅

Okay will do!

jjallaire commented:
Could you though bring the branch up to date w/ main?

jjallaire commented on src/inspect_ai/log/_transcript.py:
```diff
@@ -42,6 +43,7 @@ def __init__(self, events: list[Event]) -> None: ...
 
     def __init__(self, events: list[Event] | None = None) -> None:
         self._event_logger = None
+        self._context = WalkContext(message_cache={})
```
I don't think we need this because these `ModelEvent` instance already have shared references to their `ChatMessage` objects (the big change here is to make transcripts memory efficient after deserialization -- during the actual sample run we already have the sharing we want).

jjallaire commented:
Let's have @rasmusfaber  pick this up when he comes back (I don't think either one of us can be adequately confident we are making the right moves w/o investing a bunch of time)

sjawhar commented:
@rasmusfaber there's a merge conflict

rasmusfaber commented:
> @rasmusfaber there's a merge conflict

Resolved now.
